File: solutions/fortran/space-age/1/space_age.f90

module space_age
   implicit none
contains

   double precision function age_in_years(planet, seconds)
      character(len=*), intent(in) :: planet
      double precision, intent(in) :: seconds
      double precision :: earth_years
      earth_years = seconds/31557600.d0
      select case (planet)
      case ("Mercury")
         age_in_years = earth_years/0.2408467d0
      case ("Venus")
         age_in_years = earth_years/0.61519726d0
      case ("Earth")
         age_in_years = earth_years
      case ("Mars")
         age_in_years = earth_years/1.8808158d0
      case ("Jupiter")
         age_in_years = earth_years/11.862615d0
      case ("Saturn")
         age_in_years = earth_years/29.447498d0
      case ("Uranus")
         age_in_years = earth_years/84.016846d0
      case ("Neptune")
         age_in_years = earth_years/164.79132d0
      end select
   end function

end module

File: solutions/haskell/space-age/1/src/SpaceAge.hs
module SpaceAge (Planet (..), ageOn) where

-- import Data.List
import Data.Maybe

data Planet
  = Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune
  deriving (Eq)

planetInEarthYears :: [(Planet, Float)]
planetInEarthYears = [(Mercury, 0.2408467), (Venus, 0.61519726), (Earth, 1), (Mars, 1.8808158), (Jupiter, 11.862615), (Saturn, 29.447498), (Uranus, 84.016846), (Neptune, 164.79132)]

ageOn :: Planet -> Float -> Float
ageOn planet =
  (/ 31557600)
    . (/ fromMaybe 1 (lookup planet planetInEarthYears))

{- ageOn :: Planet -> Float -> Float
ageOn planet seconds = case find (\(pl, _) -> pl == planet) planetInEarthYears of
  Just (_, earthYears) -> seconds / 31557600 / earthYears
  _ -> 0 -}

{- findPlanet :: Planet -> Maybe (Planet, Float)
findPlanet planet = find (\(pl, _) -> pl == planet) planetInEarthYears

tupleAndSeconds :: Fractional p => Maybe (a, p) -> p -> p
tupleAndSeconds (Just (_, earthYears)) secs =
  secs / 31557600 / earthYears
tupleAndSeconds Nothing _ = 0 -}

-- ageOn :: Planet -> Float -> Float
-- ageOn = tupleAndSeconds . findPlanet